lib/constraints-store.js

```diff
@@ -0,0 +1,47 @@
+'use strict'
+
+const { assert } = require('console')
+
+function ConstraintsStore (strategies) {
+  if (!(this instanceof ConstraintsStore)) {
+    return new ConstraintsStore(strategies)
+  }
+
+  this.strategies = strategies
+}
+
+ConstraintsStore.prototype.set = function (constraints, store) {
+  // TODO: Should I check for existence of at least one constraint?
+  if (typeof constraints !== 'object' || constraints === null) {
+    throw new TypeError('Constraints should be an object')
+  }
+
+  Object.keys(constraints).forEach(kConstraint => {
+    assert(this.strategies[kConstraint] !== null, `No strategy available for handling the constraint '${kConstraint}'`)
+    this.strategies[kConstraint].set(constraints[kConstraint], { store, constraints })
+  })
+
+  return this
+}
+
+ConstraintsStore.prototype.get = function (constraints, method) {
+  if (typeof constraints !== 'object' || constraints === null) {
+    throw new TypeError('Constraints should be an object')
+  }
+
+  var returnedStore = null
+  const keys = Object.keys(constraints)
+  for (var i = 0; i < keys.length; i++) {
+    const kConstraint = keys[i]
+    assert(this.strategies[kConstraint] !== null, `No strategy available for handling the constraint '${kConstraint}'`)
+    const storedObject = this.strategies[kConstraint].get(constraints[kConstraint])
+    if (!storedObject || !storedObject.store || !storedObject.store[method]) return null
+    // TODO: Order of properties may result in inequality
+    if (JSON.stringify(constraints) !== JSON.stringify(storedObject.constraints)) return null
```

Comment:
To illustrate the purpose of that check, let's take these routes as an example: 

```
router.on('GET', '/users', { constraints: { version: '2.0.0', host: 'www.airhorns.dev' } }, handler) // Route 1
router.on('GET', '/users', { constraints: { version: '1.*.0', host: 'auth.airhorns.dev' } }, handler)   // Route 2
```

Using the constraints extracted from a request: `{ version: '1.0.0', host: 'www.airhorns.com' }`, when we call the get method of the `ConstraintsStore`, each constraint is checked separately using its defined strategy, so in this case the version constraint will match route 2, while the host constraint will match route 1. Adding this comparison allows us to avoid this edge case.

I agree that it isn't the most efficient solution, in fact now that I think of it it's wrong because the comparison will fail in non-exact match cases (in this example the version constraint has a wildcard, while the request has the exact version).

I'll try to come up with an alternative.